File: src/org/genyris/io/StringInStream.java
// Copyright 2008 Peter William Birch <birchb@genyis.org>
//
// This software may be used and distributed according to the terms
// of the Genyris License, in the file "LICENSE", incorporated herein by reference.
//
package org.genyris.io;

import org.genyris.exception.GenyrisException;
import org.genyris.interp.Environment;


public class StringInStream implements InStream {

    private final char[] _value;
    private int _readPointer;
    private int _lineCount;

    public StringInStream(String astring) {
        _value = astring.toCharArray();
        _readPointer = 0;
        _lineCount = 1;
    }

    public boolean hasData() {
        return this._readPointer <_value.length;
    }

    public char readNext() {
        if( _value[_readPointer] == '\n' ) {
            _lineCount++;
        }
        return _value[_readPointer++];
    }

    public  void unGet(char x) throws LexException {
        throw new LexException("PipeInStream: unGet not supported!");

    }

    public void close() throws GenyrisException {}

	public void resetAfterError() {
	}
	
    public void withinExpression(Environment env) {
    }

    public void beginningExpression() {
    }

    public int getLineNumber() {
        return this._lineCount;
    }
    public String getFilename() {
        return "StringInStream " + hashCode();
    }

}

File: src/org/genyris/io/StringInStream.kt
// Copyright 2008 Peter William Birch <birchb@genyis.org>
//
// This software may be used and distributed according to the terms
// of the Genyris License, in the file "LICENSE", incorporated herein by reference.
//
package org.genyris.io

import org.genyris.exception.GenyrisException
import org.genyris.interp.Environment
import java.io.CharArrayReader
import java.io.Reader

class StringInStream(astring: String) : InStream {
    private val _value: CharArray
    private var _readPointer = 0
    private var _lineCount = 1

    init {
        _value = astring.toCharArray()
    }

    @kotlin.Throws(UnsupportedOperationException::class)
    override fun getReader(): Reader {
        return CharArrayReader(_value)
    }

    override fun hasData(): Boolean {
        return this._readPointer < _value.size
    }

    override fun readNext(): Char {
        if (_value[_readPointer] == '\n') {
            _lineCount++
        }
        return _value[_readPointer++]
    }

    @kotlin.Throws(LexException::class)
    override fun unGet(x: Char) {
        throw LexException("PipeInStream: unGet not supported!")
    }

    @kotlin.Throws(GenyrisException::class)
    override fun close() {
    }

    override fun resetAfterError() {
    }

    override fun withinExpression(env: Environment?) {
    }

    override fun beginningExpression() {
    }

    override fun getLineNumber(): Int {
        return this._lineCount
    }

    override fun getFilename(): String {
        return "StringInStream " + hashCode()
    }
}
